pytorch_lightning/plugins/environments/default_environment.py

```diff
@@ -0,0 +1,58 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from typing import Optional
+
+from pytorch_lightning.plugins.environments.cluster_environment import ClusterEnvironment
+from pytorch_lightning.utilities.distributed import find_free_network_port
+
+
+class DefaultEnvironment(ClusterEnvironment):
```

Comment:
Why not let ClusterEnvironment have these default implementations instead of separating the interface `ClusterEnvironment` from `DefaultEnvironment`?

Reply:
Yes, good question. I thought about it too. The idea was to have the abstract class as an interface and by marking methods as abstract the user would know exactly what to implement. Taking this into consideration, would you still recommend merging them together?

Reply:
Notice how `master_address()` and `master_port()` are duplicate in both `DefaultEnvironment` and `TorchElasticEnvironment` because each subclasses the interface.

I'm not an expert on ABC's tricks. Is there any way we can set `ClusterEnvironment` as the abstract interface AND `DefaultEnvironment` as the default implementation? That would be ideal.

I guess the simplest way to solve it is by merging the interface and the default implementation.

Reply:
They are not exactly duplicates. I know what you mean, maybe the name DefaultEnvironment is not suitable here. 
Essentially, there is a key difference between the "DefaultEnvironment" and the others that we currently have, and that is the way processes are spawned (controlled by spawns_children bool method). Making this the superclass from which all derive would be suboptimal. Perhaps it should be called LightningEnvironment or something like that?

Reply:
BasicEnvironment?

Reply:
> They are not exactly duplicates

Why don't we make them be? Isn't their functionality the same for everything but the `world_size`, `spawns_children` and a few other methods?

Reply:
Agreed that `DefaultEnvironment` is probably not the best descriptor, `BaseEnvironment` would fit better.

Reply:
Looking into it further, I'm not so sure anymore. I guess it depends on whether users are encouraged (or only allowed) to inherit from cluster or default/base.

Reply:
As we are using localhost `127.0.0.1`, we could call it `LocalEnvironment` which might be more appropriate.

Reply:
The typical naming convention here would be to add `Abstract` to the abstract base class. In this case, you'd change `DefaultEnvironment` to `AbstractClusterEnvironment`. This quite explicitly tells users that they cannot instantiate it directly. The basic instantiable subclass would then be called `ClusterEnvironment`.  Anyone looking to provide a custom environment would then extend `AbsctractClusterEnvironment`.

Reply:
Regarding the duplicate code in `DefaultEnvironment` and `TorchElasticEnvironment`.... one option would be to make a `LocalhostEnvironment` class, which extends the abstract `ClusterEnvironment` and only implements `master_port` and `master_address` to default to `127.0.0.1`. `DefaultEnvironment` and `TorchElasticEnvironment` would then extend `LocalhostEnvironment`.

Reply:
I renamed DefaultEnvironment to LightningEnvironment for now. I want to clear up some potential points of confusion: The previous name was a poor choice, I admit that, because it would suggest that DefaultEnvironment holds the default implementation and others should inherit from that. But that is the opposite I wanted to achieve. The goal here is to provide a **specification**, and therefore the abstract class.
I do not plan to provide a base class that implements all methods, and that all environments inherit from. We need to hold back here and not abuse inheritance for sole purpose "code reuse". 
I'll see what I can do about the "duplicated code" for master address and master port.